runners/flink/src/test/java/org/apache/beam/runners/flink/PortableExecutionTest.java

```diff
@@ -0,0 +1,234 @@
+import org.junit.runners.Parameterized.Parameters;
+
+/**
+ * Tests the execution of a pipeline from specification to execution on the portable Flink runner.
+ * Exercises job invocation, executable stage translation and deployment with embedded Flink for
+ * batch and streaming.
+ */
+@RunWith(Parameterized.class)
+public class PortableExecutionTest implements Serializable {
+
+  @Parameters
+  public static Object[] data() {
+    return new Object[] {true, false};
+  }
+
+  @Parameter public boolean isStreaming;
+
+  private transient ExecutorService sdkHarnessExecutor;
+  private transient Future<?> sdkHarnessExecutorFuture;
+
+  private DockerJobBundleFactory createJobBundleFactory(JobInfo jobInfo) throws Exception {
+    return new DockerJobBundleFactory(jobInfo) {
+
+      @Override
+      protected ServerFactory getServerFactory() {
+        return ServerFactory.createWithUrlFactory(
+            (host, port) -> HostAndPort.fromParts("localhost", port).toString());
+        //return InProcessServerFactory.create();
+      }
+
+      @Override
+      protected EnvironmentFactory getEnvironmentFactory(
+          GrpcFnServer<FnApiControlClientPoolService> controlServer,
+          GrpcFnServer<GrpcLoggingService> loggingServer,
+          GrpcFnServer<ArtifactRetrievalService> retrievalServer,
+          GrpcFnServer<StaticGrpcProvisionService> provisioningServiceServer,
+          Source clientSource,
+          IdGenerator idGenerator) {
+        String workerId = "fakeWorkerId";
+        ThreadFactory threadFactory = new ThreadFactoryBuilder().setDaemon(true).build();
+        // Create the SDK harness, and wait until it connects
+        sdkHarnessExecutor = Executors.newSingleThreadExecutor(threadFactory);
+        sdkHarnessExecutorFuture =
+            sdkHarnessExecutor.submit(
+                () -> {
+                  try {
+                    FnHarness.main(
+                        workerId,
+                        PipelineOptionsFactory.create(),
+                        loggingServer.getApiServiceDescriptor(),
+                        controlServer.getApiServiceDescriptor(),
+                        ManagedChannelFactory.createDefault(),
+                        //InProcessManagedChannelFactory.create(),
+                        HarnessStreamObserverFactories.fromOptions(
+                            PipelineOptionsFactory.create()));
+                  } catch (Exception e) {
+                    throw new RuntimeException(e);
+                  }
+                });
+        try {
+          InstructionRequestHandler controlClient =
+              clientSource.take(workerId, Duration.ofSeconds(2));
+          return (container) -> RemoteEnvironment.forHandler(container, controlClient);
+        } catch (Exception ex) {
+          throw new RuntimeException(ex);
+        }
+      }
+    };
+  }
+
+  @Before
+  public void setup() {
+    DockerJobBundleFactory.factory = this::createJobBundleFactory;
```

Comment:
This seems to be a bit hacky. Shall we expose the factory method explicitly.

Reply:
This is a test internal that does not need to be used anywhere else.

Reply:
As mentioned below, this could also be done using a service loader with the resource in test scope, my focus in this PR is more to get the basic test coverage in place.